clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py

```diff
@@ -315,54 +372,79 @@ def _generate_config(self, file_path, **kwargs):
-            self.logger.warning(msg)
-            for addr in node.status.addresses:
-                if addr.type == "InternalIP":
-                    external_node_hosts.append(addr.address)
-
-        if len(external_node_hosts) == 0:
-            msg = "Error connecting to Kubernetes cluster. No external node addresses found. You may pass in KubernetesContainerManager(useInternalIP=True) to connect to local Kubernetes cluster"
-            self.logger.error(msg)
-            raise ClipperException(msg)
-
-        self.external_node_hosts = external_node_hosts
-        self.logger.info("Found {num_nodes} nodes: {nodes}".format(
-            num_nodes=len(external_node_hosts),
-            nodes=", ".join(external_node_hosts)))
+        if any(v in [CLUSTER_IP, NODE_PORT] for v in self.service_types.values()):
+            nodes = self._k8s_v1.list_node()
+
+            external_node_hosts = []
+            for node in nodes.items:
+                for addr in node.status.addresses:
+                    if addr.type == "ExternalDNS":
+                        external_node_hosts.append(addr.address)
+
+            if len(external_node_hosts) == 0 and self.useInternalIP:
+                msg = "No external node addresses found. Using Internal IP address"
+                self.logger.warning(msg)
+                for addr in node.status.addresses:
+                    if addr.type == "InternalIP":
+                        external_node_hosts.append(addr.address)
+
+            if len(external_node_hosts) == 0:
+                msg = "Error connecting to Kubernetes cluster. No external node addresses found. You may pass in KubernetesContainerManager(useInternalIP=True) to connect to local Kubernetes cluster"
+                self.logger.error(msg)
+                raise ClipperException(msg)
+
+            self.external_node_hosts = external_node_hosts
+            self.logger.info("Found {num_nodes} nodes: {nodes}".format(
+                num_nodes=len(external_node_hosts),
+                nodes=", ".join(external_node_hosts)))
 
         try:
-            mgmt_frontend_ports = self._k8s_v1.read_namespaced_service(
+            v1service = self._k8s_v1.read_namespaced_service(
                 name="mgmt-frontend-at-{cluster_name}".format(
                     cluster_name=self.cluster_name),
-                namespace=self.k8s_namespace).spec.ports
+                namespace=self.k8s_namespace)
+            mgmt_frontend_ports = v1service.spec.ports
             for p in mgmt_frontend_ports:
-                if p.name == "1338":
+                if int(p.name) == CLIPPER_INTERNAL_MANAGEMENT_PORT:
                     self.clipper_management_port = p.node_port
-                    self.logger.info("Setting Clipper mgmt port to {}".format(
-                        self.clipper_management_port))
 
-            query_frontend_ports = self._k8s_v1.read_namespaced_service(
+            if self.service_types['management'] in [CLUSTER_IP, NODE_PORT]:
```

Comment:
Why don't we read service_types from our Kubernetes definitions (yaml file) in the `connect` function? Otherwise we should always specify the correct `self.service_types` whenever we initiate `KubernetesContainerManager`. 

It shouldn't matter when we initiate `KubernetesContainerManager` for the first time for `start_clipper`. But, let's say we already started clipper and just wanted to connect to the cluster using `ClipperConnection(KubernetesContainerManager()).connect()`. If you forget to write `service_types`, it will be set to be None and this part might be broken.

Reply:
@rkooo567 You're right. Thank you for your kindly advice. I will fix it as soon as possible.

Reply:
@rkooo567 : According to your advice, I added a new mechanism to read/write `service_types` from/to a yaml file. You can connect to the executing Clipper cluster easily with `service_types`(None), which will be loaded automatically from a yaml file saved when we started it. If you call 'stop_all()', the yaml file will be removed.

Reply:
Thank you! Everything looks good to me :)